### PR TITLE
[Frontend] Speaker embedding support for speech and voices APIs

### PR DESCRIPTION
**Pull request by marksverdhei**
## Summary

Add `speaker_embedding` support across the Qwen3-TTS speech API — both as a
per-request parameter on `/v1/audio/speech` and as a new input source on
`POST /v1/audio/voices`.

The speaker encoder extracts a fixed-size embedding from reference audio on
every request. For repeated use of the same voice, this is redundant.
Pre-computed embeddings also enable voice blending via SLERP interpolation
and offline batch processing.

Complements #1201 (voice upload API): upload provides persistent server-side
storage from audio; embedding passthrough provides a stateless path for
ephemeral or synthetic embeddings.

### Changes

**Per-request embedding (`/v1/audio/speech`):**
- `speaker_embedding` field on request (1024-dim for 0.6B, 2048-dim for 1.7B)
- Mutual exclusivity with `ref_audio`, finiteness validation, `max_length=8192`
- Auto-implies `x_vector_only_mode=True`
- Also added to `StreamingSpeechSessionConfig` for WebSocket parity

**Named voice registration (`/v1/audio/voices`):**
- `POST` accepts optional `speaker_embedding` (JSON float list) as
  alternative to `audio_sample` — mutually exclusive
- Stored as safetensors, marked `cache_status: "ready"` immediately
- `GET` response includes `embedding_source: "audio" | "direct"` and
  `embedding_dim`

**Model layer:**
- Fix `_as_singleton` destructive unwrap for list data from IPC
- Handle both tensor and list embedding formats in `_build_prompt_embeds`

**Examples:**
- `--speaker-embedding` flag in `openai_speech_client.py` (loads JSON)
- `speaker_embedding_interpolation.py`: ECAPA-TDNN extraction + SLERP
  interpolation CLI tool

## Test plan

- [x] Mutual exclusivity: `speaker_embedding` + `ref_audio` rejected at
  pydantic parse level
- [x] NaN/Inf values rejected at parse level
- [x] `speaker_embedding` implies `x_vector_only_mode` (no `ref_text` required)
- [x] Edge cases: empty list, non-standard dims (warning only)
- [x] `_build_tts_params` produces correct `voice_clone_prompt` structure
- [x] Embedding upload via voices endpoint, appears in listing with correct
  source and dimension
- [x] Audio + embedding mutually exclusive on voices endpoint
- [x] Invalid JSON / NaN values rejected on voices endpoint (400)
- [x] SLERP math: endpoints, midpoint norm, parallel/antiparallel edge cases
- [x] E2E tests across 0.6B-Base, 0.6B-CustomVoice, 1.7B-Base models

Signed-off-by: marksverdhei <marksverdhei@hotmail.com>

### CODE REVIEW

File: vllm_omni/model_executor/models/qwen3_tts/qwen3_tts.py [hunk 1]
```diff
@@ -793,6 +794,18 @@ def generate_voice_clone(
 
         self._validate_languages(languages)
 
+        # If a pre-computed speaker embedding is provided, build prompt items directly
+        if speaker_embedding is not None and voice_clone_prompt is None:
+            spk_tensor = torch.tensor(speaker_embedding, dtype=torch.float32).to(self.device)
```
[the remark this answers is not in the document]
Comment (marksverdhei) [reply vllm_omni/model_executor/models/qwen3_tts/qwen3_tts.py]: > *This reply was generated by an AI agent (Claude Code) acting on behalf of @marksverdhei.*

Addressed in 6cdea6e. Removed the hardcoded `dtype=torch.float32` — the tensor is now created with default dtype and the downstream `generate_speaker_prompt()` handles conversion to `self.talker.dtype` (bfloat16) before injection into the codec input embeddings.

File: vllm_omni/entrypoints/openai/serving_speech.py [hunk 2]
```diff
@@ -93,12 +93,21 @@ def _validate_tts_request(self, request: OpenAICreateSpeechRequest) -> str | Non
             if self.supported_speakers and request.voice not in self.supported_speakers:
                 return f"Invalid speaker '{request.voice}'. Supported: {', '.join(sorted(self.supported_speakers))}"
 
+        # Validate speaker_embedding constraints
+        if request.speaker_embedding is not None:
+            if task_type != "Base":
+                return "'speaker_embedding' is only valid for Base task"
+            if request.ref_audio is not None:
+                return "'speaker_embedding' and 'ref_audio' are mutually exclusive"
```
[the remark this answers is not in the document]
Comment (marksverdhei) [reply vllm_omni/entrypoints/openai/serving_speech.py]: > *This reply was generated by an AI agent (Claude Code) acting on behalf of @marksverdhei.*

Addressed in 6cdea6e. Added bounds validation (64–8192) for the embedding length. The exact dimension is model-dependent (1024 for 0.6B, 2048 for 1.7B), so we validate against a reasonable range rather than a fixed size. This catches obviously malformed vectors with a clear client-facing error before they reach model execution.

**Comment by linyueqian** [comment]
Heads up — PR #1201 adds a voice upload API (`POST /v1/audio/voices`) and modifies the same validation/param-building code in `serving_speech.py`. To avoid conflicts, could you wait for #1201 to merge first, then rebase this PR on top?

After rebasing you'd need to:
- Unify the Base task validation to handle all three voice sources (`ref_audio`, `speaker_embedding`, and uploaded voices)
- Align `_build_tts_params` with #1201's uploaded voice handling
- Consider also accepting `speaker_embedding` in `POST /v1/audio/voices` for registering embeddings as named voices

**Comment by hsliuustc0106** [comment]
@vllm-omni-reviewer

File: examples/online_serving/qwen3_tts/speaker_embedding_interpolation.py
Comment (marksverdhei) [inline examples/online_serving/qwen3_tts/speaker_embedding_interpolation.py:1]: Remove

File: examples/online_serving/qwen3_tts/openai_speech_client.py
Comment (marksverdhei) [inline examples/online_serving/qwen3_tts/openai_speech_client.py:1]: Remove? Should we prefer safetensors over json here?

File: vllm_omni/entrypoints/openai/protocol/audio.py
Comment (marksverdhei) [inline vllm_omni/entrypoints/openai/protocol/audio.py:1]: Can be 2048 dim not just 1024

**Review by marksverdhei — COMMENTED** [review]
Just spotted a couple todos

File: vllm_omni/entrypoints/openai/protocol/audio.py [hunk 3]
```diff
@@ -44,6 +44,12 @@ class OpenAICreateSpeechRequest(BaseModel):
         default=None,
         description="Use speaker embedding only without in-context learning (Base task)",
     )
+    speaker_embedding: list[float] | None = Field(
+        default=None,
+        description="Pre-computed speaker embedding vector (1024-dim). "
```
Comment (marksverdhei) [inline vllm_omni/entrypoints/openai/protocol/audio.py]:
```suggestion
        description="Pre-computed speaker embedding vector"
```

**Comment by marksverdhei** [comment]
I'm gonna follow up on this now but a few things needed before ready for review. I'm unsure exactly how to land on the generic-specific trade-off as for now but i'll come up with something

**Comment by JuanPZuluaga** [comment]
Hello @marksverdhei, thanks for this work. do you plan to work/merge it anytime soon? I find it very useful and can help to debug/work on it.

**Comment by marksverdhei** [comment]
> Hello @marksverdhei, thanks for this work. do you plan to work/merge it anytime soon? I find it very useful and can help to debug/work on it.

Hi, sorry for late reply. i will try to get it ready for review today.,

File: vllm_omni/entrypoints/openai/serving_speech.py [hunk 4]
```diff
@@ -756,7 +869,18 @@ def _build_tts_params(self, request: OpenAICreateSpeechRequest) -> dict[str, Any
...
+            # Store as plain float list (not tensor) so it survives msgspec
+            # serialization through the EngineCore IPC boundary.  The talker's
+            # _build_prompt_embeds converts it back to a tensor on the GPU.
+            params["voice_clone_prompt"] = [
+                {
+                    "ref_spk_embedding": list(request.speaker_embedding),
+                }
+            ]
+            # speaker_embedding implies x_vector_only_mode
+            params["x_vector_only_mode"] = [True]
```
[the remark this answers is not in the document]
Comment (marksverdhei) [reply vllm_omni/entrypoints/openai/serving_speech.py:956]: Addressed in 305c3d6. `_validate_tts_request()` now sets `request.x_vector_only_mode = True` immediately after the speaker_embedding constraint checks, before the Base task ref_text validation runs. Callers no longer need to explicitly pass `x_vector_only_mode` when using `speaker_embedding`.

File: vllm_omni/entrypoints/openai/protocol/audio.py [hunk 5]
```diff
@@ -121,6 +145,11 @@ class StreamingSpeechSessionConfig(BaseModel):
     ref_audio: str | None = None
     ref_text: str | None = None
     x_vector_only_mode: bool | None = None
+    speaker_embedding: list[float] | None = Field(
+        default=None,
+        max_length=_MAX_EMBEDDING_DIM,
+        description="Pre-computed speaker embedding vector. Mutually exclusive with ref_audio.",
+    )
```
[the remark this answers is not in the document]
Comment (marksverdhei) [reply vllm_omni/entrypoints/openai/protocol/audio.py:208]: Addressed in 305c3d6. `_generate_and_send()` in `serving_speech_stream.py` now forwards `config.speaker_embedding` when constructing the `OpenAICreateSpeechRequest`.
